### ISSUE TITLE
No Upsampling2D layer in Tensorflow C++

### ISSUE THREAD

**Issue by mortalrahu** [tensorflow/tensorflow#38612]
I am Rahul, an M.Sc. in Digital Engineering student at University of Magdeburg, Germany. I am currently pursuing my master thesis and it involves developing a U-Net (a neural network designed for biomedical image segmentation, you can read more about it here : https://arxiv.org/pdf/1505.04597.pdf). And deploying/loading it in a C++ Application. The U-Net has a couple of layers namely Concatenate, UpSampling2D which are not common in Image processing. 

To deploy it I am guessing I need all the layers used U-Net available in Tensorflow C++ here : https://www.tensorflow.org/api_docs/cc/group/nn-ops

I can't find Upsampling2D there. 

Am I looking at the wrong location? 

I would like to know if  loading the protobuf and predicting using it is independent of the structure and layers used in the model I developed or it only supports standard layers such as flatten, convolution, maxpool etc….

**Comment by Leslie-Fang** [comment]
@mortalrahu There is a workaround I used to use which combines the API of 
```
tf.transpose 
tf.image.resize
```
you can take a try

**Comment by Saduf2019** [comment]
@mortalrahu
please update on the above comment

**Comment by mortalrahu** [comment]
@Leslie-Fang 

Thanks for the response.

Are there any open source projects that are completely built from scratch in Tensorflow C++ , meaning trained and working completely in C++.  (Especially if it has the Upsampling alternative usage you suggest) 

or projects where módel is trained in python, frozen, saved and then loaded in C++ to do predictions

**Comment by Leslie-Fang** [comment]
@mortalrahu Sorry, haven't used the C++ API before,

**Comment by alextp** [comment]
Upsampling2d is just tf.image.resize; it's implemented using a bunch of different ops depending on what type of resizing you want to do (see the code in https://github.com/tensorflow/tensorflow/blob/fadfcdf27ba3973525b7ef24f5d2cb10a13231bd/tensorflow/python/ops/image_ops_impl.py#L1316 ); those individual ops are all available through the TF C++ API.

We do recommend though that you use python to build your model and C++ to execute it, as taking gradients and doing optimization in the TF C++ API directly doesn't work very well right now.
